dask_kubernetes/core.py

```diff
@@ -415,7 +413,104 @@ def __init__(
         self.kwargs = kwargs
         super().__init__(**self.kwargs)
 
-    def _get_pod_template(self, pod_template, pod_type):
+    @property
+    def pod_template(self):
+        return self.__pod_template
+
+    @pod_template.setter
+    def pod_template(self, pod_template):
+        if isinstance(pod_template, str):
+            msg = (
+                f"Expected a kubernetes.client.V1Pod object, got '{pod_template}'. "
+                f"If trying to pass a yaml filename then use "
+                f"{self.__class__.__name__}.from_yaml"
+            )
+            raise TypeError(msg)
+
+        if isinstance(pod_template, dict):
+            msg = (
+                f"Expected a kubernetes.client.V1Pod object, got {pod_template}. "
+                f"If trying to pass a dictionary specification then use "
+                f"{self.__class__.__name__}.from_dict"
+            )
+            raise TypeError(msg)
+        self.__pod_template = pod_template
+
+    @property
+    def _generate_name(self):
+        _generate_name = self.__generate_name or dask.config.get("kubernetes.name")
+        _generate_name = _generate_name.format(
+            user=getpass.getuser(), uuid=str(uuid.uuid4())[:10], **os.environ
+        )
+        return escape(_generate_name)
+
+    @property
+    def _namespace(self):
+        return (
+            self.__namespace
+            or dask.config.get("kubernetes.namespace")
+            or _namespace_default()
+        )
+
+    @property
+    def name(self):
+        return self._generate_name
+
+    @property
+    def namespace(self):
+        return self._namespace
+
+    @property
+    def _n_workers(self):
+        return (
+            self.n_workers
+            if self.n_workers is not None
+            else dask.config.get("kubernetes.count.start")
+        )
+
+    @property
+    def _idle_timeout(self):
+        return self.idle_timeout or dask.config.get("kubernetes.idle-timeout")
```

Comment:
Moving these attributes to properties seems to add a large amount of [indirection](https://matthewrocklin.com/blog/work/2019/06/23/avoid-indirection), re-execution and complexity here. What is the reasoning for this?

Reply:
- python properties effectively make (parts of) the code *stateless* and *functional*. Anything stateful is harder to unit test, debug and maintain than pure functional code. IMHO the *state* of the object should be described by the __init__ kwargs (or as close as possible) whenever possible.

- if re-execution performance becomes an issue, property getters can easily be cached with `lru_cache` (or an explicit internal `__idle_timeout` that's set on first access).

While it adds a little boilerplate, it allows us to (for example) add a unit test to check that `"kubernetes.idle-timeout"` is used as default *in isolation from other code paths*

Reply:
I would say that IMHO it falls under ["functions are a good idea"](https://matthewrocklin.com/blog/work/2019/06/23/avoid-indirection), maybe less relevant for the specific `_idle_timeout` getter, but *definitely* the case for the `rendered_worker_pod_template` property

Reply:
That sounds like something that should be discussed as an issue and then implemented separately as a PR if the community agrees with your logic.  In the meantime I recommend that you revert these changes so that the main part of this PR isn't blocked on them.

Reply:
@mrocklin having these attributes available as properties outside `_start()` is integral to the PR, so I am open to alternative suggestions but I don't see an easy way to revert a sub-set of changes?